docs: Add comments to Blog (#1753)
tendermint/starport: docs/guide/blog/comment-blog.md

---
description: Add comment on Blog
order: 2
---

# Add comments to Blog

> In this tutorial, you will create a new message module called comment. The module will let you read and write comments to the existing blog blockchain.

Reviewer: ```suggestion
In this tutorial, you will create a new message to add a comment to a blog post. You will be implementing logic for writing comments to the blockchain as well as querying for blog posts with associated comments.
```

> You can only add comments to a post which is no older than 100 Blocks. 

Reviewer: ```suggestion
You can only add comments to a post that is no older than 100 blocks. 
```

**Note:** This value has been hard coded to a low number for rapid testing. You can increase it to a greater number to achieve longer period of time before commenting is stopped.

### Prerequisites:

- This tutorial is an extension of previously written blog tutorial. Make sure you complete that first before proceeding with this tutorial.
- This tutorial also assumes basic knowledge of blog tutorial implementation.
- Make sure you are inside the `blog` directory created in the previous blog tutorial.

## Create a new message called comment

> To create a new message module for comment, use the `message` command:

Reviewer: There is no "message module".

Author: I'll rephrase as: To create a new message, use the `message` command:

Reviewer: That works.

```bash

> starport scaffold message create-comment blogID:uint title body

Reviewer: ```suggestion
starport scaffold message create-comment postID:uint title body
```

Author: Since from this point onwards there are major changes suggested.. I will re work on the PR.

		//...
		case *types.MsgCreateComment:
			res, err := msgServer.CreateComment(sdk.WrapSDKContext(ctx), msg)
			return sdk.WrapServiceResult(ctx, res, err)
			//...
		}
	}
}
```

The `case *types.MsgCreateComment` statement handles messages of type `MsgCreateComment`, calls the `CreateComment` method, and returns back the response.

## Process Messages

In the newly scaffolded `x/blog/keeper/msg_server_create_comment.go` file, you can see a placeholder implementation of the `CreateComment` function. Right now it does nothing and returns an empty response. For your blog chain, you want the contents of the message (title and body) to be written to the state as a new comment.

You need to do two things:

- Create a variable of type `Comment` with title and body from the message
- Check if the the comment posted for the respective blog id exists and comment is not older than 100 blocks.
- Append this `Comment` to the store

```go
import (
    //...
    sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
)

func (k msgServer) CreateComment(goCtx context.Context, msg *types.MsgCreateComment) (*types.MsgCreateCommentResponse, error) {
	ctx := sdk.UnwrapSDKContext(goCtx)

> 	BlockHeight := CreatedAt() // Invoke method to get the block height of Post

Reviewer: Use `k.GetPost(ctx, msg.PostId)` to get a post and get the height of creation from the post object. Store it in a local variable.

> 	PostID := CreatedId() // Invoke method to get the Post ID

Reviewer: You already have the post ID inside `msg`.

	// Create variable of type comment
	var comment = types.Comment{
		Creator: msg.Creator,
		Id:      msg.Id,
		Body:    msg.Body,
		Title:   msg.Title,
		BlogID:  msg.BlogID,
		CreatedAt: ctx.BlockHeight(),
	}

	// Check if the Post Exists for which a comment is being created

> 	if comment.BlogID > PostID {

Reviewer: This error should be handled above, when `GetPost` returns an error, you throw an error that a blog post with a given post ID can't be found.

		return nil, sdkerrors.Wrapf(types.ErrID, "Post Blog Id %d does not exist for which comment with Blog Id %d was made", PostID, comment.BlogID)
	}

	BlockHeight = BlockHeight + 100 // Hardcoded value to 100. This can be changed as per requirement.
	
	// Check if the comment is older than the Post. If more than 100 blocks, then return error.
	if comment.CreatedAt > BlockHeight {
		return nil, sdkerrors.Wrapf(types.ErrCommentOld, "Comment created at %d is older than post created at %d", comment.CreatedAt, BlockHeight)

> 	} else {

Reviewer: Don't need an `else`. `if` case returns, otherwise the logic continues.

		id := k.AppendComment(ctx, comment)
		return &types.MsgCreateCommentResponse{Id: id}, nil
	}
}
```

When Comment's validity is checked, it throws 2 error messages - `ErrID` and `ErrCommendOld`. You can define the error messaged by adding it to errors definition in x/blog/types/errors.go

```go
//...
var (
	ErrCommentOld = sdkerrors.Register(ModuleName, 1300, "")
)

var (
	ErrID = sdkerrors.Register(ModuleName, 1400, "")
)
```

Notice, 2 methods `CreatedAt` and `CreatedId` are being invoked inside `msg_server_create_comment.go`. They are responsible for returning the Block height and latest ID of Post respectively.

Define `CreatedAt` and `CreatedId` method inside `x/blog/keeper/msg_server_create_post.go`

```go
//...

> var BlockHeight int64

Reviewer: Please, don't use globals.

var BlockId uint64 

func (k msgServer) CreatePost(goCtx context.Context, msg *types.MsgCreatePost) (*types.MsgCreatePostResponse, error) {
  //...
  var post = types.Post{
    //...
    CreatedAt: ctx.BlockHeight(),
  }

    // Define the globally declared `BlockHeight` variable
    BlockHeight = post.CreatedAt

    id := k.AppendPost(ctx, post)
    //...

    // After the post has been appended, fetch the latest id in globally declared `BlockId` variable
    BlockId = id
}

// Define two functions to return BlockId and Height respectively.

func CreatedAt() int64 {
	return BlockHeight
}

func CreatedId() uint64 {
	return BlockId
}
```

## Write Data to the Store

Define the `Comment` type and the `AppendComment` keeper method.

When you define the `Comment` type in a proto file, Starport (with the help of `protoc`) takes care of generating the required Go files.

> Create the `proto/blog/comment.proto` file and define the `Comment` message:

Reviewer: Use `starport scaffold type` instead.

Author: If I understand correctly the ask is to use - `starport scaffold type comment`?

I will move this part to the top when we do scaffolding for message. Makes sense?

Author: However `starport scaffold type comment` throws - "component type with name comment is already created (type MsgCreateComment exists)" 
Not sure if this is what you meant?

Author: @fadeev - I think it should be first scaffolding type, and then scaffolding message. That works - Thanks!

Reviewer: > However starport scaffold type comment throws - "component type with name comment is already created (type MsgCreateComment exists)"

This is probably a bug. I'll raise an issue about this.

)

func (k Keeper) Comments(c context.Context, req *types.QueryCommentsRequest) (*types.QueryCommentsResponse, error) {
	// Throw an error if request is nil
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "invalid request")
	}
	// Define a variable that will store a list of posts
	var comments []*types.Comment
	// Get context with the information about the environment
	ctx := sdk.UnwrapSDKContext(c)
	// Get the key-value module store using the store key (in our case store key is "chain")
	store := ctx.KVStore(k.storeKey)
	// Get the part of the store that keeps posts (using post key, which is "Post-value-")
	commentStore := prefix.NewStore(store, []byte(types.CommentKey))
	// Paginate the posts store based on PageRequest
	pageRes, err := query.Paginate(commentStore, req.Pagination, func(key []byte, value []byte) error {
		var comment types.Comment
		if err := k.cdc.Unmarshal(value, &comment); err != nil {
			return err
		}
		comments = append(comments, &comment)
		return nil
	})
	// Throw an error if pagination failed
	if err != nil {
		return nil, status.Error(codes.Internal, err.Error())
	}
	// Return a struct containing a list of posts and pagination info

> 	return &types.QueryCommentsResponse{Comment: comments, Pagination: pageRes}, nil

Reviewer: The whole point of this tutorial is to write a query that returns a blog post by ID with associated comments.

	}

	// Check if the comment is older than the Post. If more than 100 blocks, then return error.
	if comment.CreatedAt > post.CreatedAt+100 {
		return nil, sdkerrors.Wrapf(types.ErrCommentOld, "Comment created at %d is older than post created at %d", comment.CreatedAt, post.CreatedAt)
	}

	id := k.AppendComment(ctx, comment)
	return &types.MsgCreateCommentResponse{Id: id}, nil
}
```

When the Comment validity is checked, it throws 2 error messages - `ErrID` and `ErrCommendOld`. You can define the error messages by adding error definitions to `x/blog/types/errors.go`:

```go
//...
var (
	ErrCommentOld = sdkerrors.Register(ModuleName, 1300, "")
	ErrID = sdkerrors.Register(ModuleName, 1400, "")
)
```

In the existing `x/blog/keeper/msg_server_create_post.go` file, you need to make a modification to add `createdAt`

```go
func (k msgServer) CreatePost(goCtx context.Context, msg *types.MsgCreatePost) (*types.MsgCreatePostResponse, error) {
	// Get the context
	ctx := sdk.UnwrapSDKContext(goCtx)
	// Create variable of type Post
	var post = types.Post{
		Creator:   msg.Creator,
		Id:        msg.Id,
		Title:     msg.Title,
		Body:      msg.Body,

> 		CreatedAt: ctx.BlockHeight(),

Reviewer: ```
error while running command go: # github.com/cosmonaut/blog/x/blog/keeper
../../x/blog/keeper/msg_server_create_comment.go:24:3: cannot use ctx.BlockHeight() (type int64) as type uint64 in field value
```

When scaffolding a comment in the following command:

```
starport scaffold list comment --no-message creator:string title:string body:string postID:uint createdAt:int 
```

`createdAt` was specified as `int`, which is a shorthand for `int32`. `ctx.BlockHeight()` is `uint64`.

Author: Made changes, `starport scaffold list comment --no-message creator:string title:string body:string postID:uint createdAt:uint`

Testing again for consistency.

```bash
blogd tx blog create-post Uno "This is the first post" --from alice
```

As before, you are prompted to confirm the transaction:

```bash
"body":{"messages":[{"@type":"/cosmonaut.blog.blog.MsgCreatePost","creator":"cosmos1dad8xvsj3dse928r52yayygghwvsggvzlm730p","title":"foo","body":"bar"}],"memo":"","timeout_height":"0","extension_options":[],"non_critical_extension_options":[]},"auth_info":{"signer_infos":[],"fee":{"amount":[],"gas_limit":"200000","payer":"","granter":""}},"signatures":[]}

confirm transaction before signing and broadcasting [y/N]: y
```

The transaction output:

```bash
{"height":"6861","txhash":"6086372860704F5F88F4D0A3CF23523CF6DAD2F637E4068B92582E3BB13800DA","codespace":"","code":0,"data":"0A100A0A437265617465506F737412020801","raw_log":"[{\"events\":[{\"type\":\"message\",\"attributes\":[{\"key\":\"action\",\"value\":\"CreatePost\"}]}]}]","logs":[{"msg_index":0,"log":"","events":[{"type":"message","attributes":[{"key":"action","value":"CreatePost"}]}]}],"info":"","gas_wanted":"200000","gas_used":"44674","tx":null,"timestamp":""}
```

Create a comment:

```bash
blogd tx blog create-comment 0  Uno "This is the first comment" --from alice
```

```bash

> {"body":{"messages":[{"@type":"/cosmonaut.blog.blog.MsgCreateComment","creator":"cosmos17pvwgu36fu37j8y9gc4pasxsj3p26ghmlqvngd","id":"0","title":"Uno","body":"This is the first comment","blogID":"2","createdAt":"0"}],"memo":"","timeout_height":"0","extension_options":[],"non_critical_extension_options":[]},"auth_info":{"signer_infos":[],"fee":{"amount":[],"gas_limit":"200000","payer":"","granter":""}},"signatures":[]}

Reviewer 2: @alijnmerchant21 we want `postID` or is `blogID` accurate in this instance?

timestamp: ""
tx: null
txhash: 0CAFC113D1C73BC0210EFEA5964EBD2EB530311169FB442C5CBF0B5E92521C41
```

## Display post and comment

```bash
blogd q blog comments 0
```

The results are output:

```bash
Comment:
- body: Let us add random comment
  createdAt: "14094"
  creator: cosmos1g7x9cpj6w0jklshe3se57tlwydx6yfl8ex5g7n
  id: "0"
  postID: "0"
  title: comment
Post:
  body: This is the first post
  createdAt: "14046"
  creator: cosmos1g7x9cpj6w0jklshe3se57tlwydx6yfl8ex5g7n
  id: "0"
  title: Uno
```

## Delete comment

```bash
blogd tx blog delete-comment 0 0 --from alice -y
```

The results are output:

```bash
code: 0
codespace: ""
data: 0A270A252F636F736D6F6E6175742E626C6F672E626C6F672E4D736744656C657465436F6D6D656E74
gas_used: "40125"
gas_wanted: "200000"
height: "253"
info: ""
logs:
- events:
  - attributes:
    - key: action
      value: DeleteComment
    type: message
  log: ""
  msg_index: 0
raw_log: '[{"events":[{"type":"message","attributes":[{"key":"action","value":"DeleteComment"}]}]}]'
timestamp: ""
tx: null
txhash: 0312234CBB9EEA1A59D474496E100AFC5A460A0E60E7D009D3E9417530148A75
```

> ## Display

Reviewer 2: @alijnmerchant21  can we give this heading a more descriptive name? What are we displaying?

create x/blog/keeper/comment_test.go
create x/blog/keeper/grpc_query_comment.go
create x/blog/keeper/grpc_query_comment_test.go
modify x/blog/module.go
modify x/blog/types/genesis.go
modify x/blog/types/genesis_test.go
modify x/blog/types/keys.go

🎉 comment added.
```

## Add a Comment to a Post

To create a new message that adds a comment to the existing post, run:

```bash
starport scaffold message create-comment postID:uint title body
```

The `starport scaffold message` command accepts `postID` and a list of fields as arguments. The fields are `title` and `body`.

Here, `postID` is the reference to previously created blog post.

The `message` command has created and modified several files:

```text
modify proto/blog/tx.proto
modify x/blog/client/cli/tx.go
create x/blog/client/cli/tx_create_comment.go
modify x/blog/handler.go
create x/blog/keeper/msg_server_create_comment.go
modify x/blog/types/codec.go
create x/blog/types/message_create_comment.go
create x/blog/types/message_create_comment_test.go

🎉 Created a message `create-comment`.
```

As always, start your development with a proto file. 

In the `proto/blog/tx.proto` file, edit `MsgCreateComment` to:

> - Add `createdAt`

Reviewer: We're saying "Add `createdAt`", but there is no `createdAt` in the code block below.

Reviewer: Also, if we're adding a field to a newly scaffolded message, shouldn't we just have `createdAt` in the scaffolding command?

- Define the `id` for `message MsgCreateCommentResponse`:

```go
message MsgCreateComment {
  string creator = 1;

>   int32 blogID = 2;

Reviewer: ```suggestion
  int32 postID = 2;
```

  string title = 3;
  string body = 4;

>   uint64 id = 5;

Reviewer: Where did `uint64 id = 5;` came from? We didn't mention it in the list above.

}

message MsgCreateCommentResponse {
  uint64 id = 1;
}
```

>  You see in the `proto/blog/tx.proto` file that the `MsgCreateComment` has five fields: creator, title, body, blogID, and id. Since the purpose of the `MsgCreateComment` message is to create new comments in the store, the only thing the message needs to return is an ID of a created comments. The `CreateComment` rpc was already added to the `Msg` service:

Reviewer: Where is `blogID` coming from?

Author: Yeah..  seems like it should be `postID`. Searched the PR further I see `postID` being used everywhere.

Reviewer 2: @alijnmerchant21 did we want `postID` here? 
```suggestion
 You see in the `proto/blog/tx.proto` file that the `MsgCreateComment` has five fields: creator, title, body, blogID, and id. Since the purpose of the `MsgCreateComment` message is to create new comments in the store, the only thing the message needs to return is an ID of a created comments. The `CreateComment` rpc was already added to the `Msg` service:
```

	ctx := sdk.UnwrapSDKContext(goCtx)

	post := k.GetPost(ctx, msg.PostID)
	postId := post.Id

	// Create variable of type comment
	var comment = types.Comment{
		Creator:   msg.Creator,
		Id:        msg.Id,
		Body:      msg.Body,
		Title:     msg.Title,
		PostID:    msg.PostID,
		CreatedAt: ctx.BlockHeight(),
	}

	// Check if the Post Exists for which a comment is being created
	if msg.PostID > postId {
		return nil, sdkerrors.Wrapf(types.ErrID, "Post Blog Id does not exist for which comment with Blog Id %d was made", msg.PostID)
	}

	// Check if the comment is older than the Post. If more than 100 blocks, then return error.
	if comment.CreatedAt > post.CreatedAt+100 {
		return nil, sdkerrors.Wrapf(types.ErrCommentOld, "Comment created at %d is older than post created at %d", comment.CreatedAt, post.CreatedAt)
	}

	id := k.AppendComment(ctx, comment)
	return &types.MsgCreateCommentResponse{Id: id}, nil
}
```

When Comment's validity is checked, it throws 2 error messages - `ErrID` and `ErrCommendOld`. You can define the error messages by adding error definitions to `x/blog/types/errors.go`:

```go
//...
var (

> 	ErrCommentOld = sdkerrors.Register(ModuleName, 1300, "")

Reviewer: Let's put both `ErrCommentOld` and `ErrID` in the same `var (...)`.

- Delete the comment from the store.

```go
package keeper

 import (
	"context"
	"encoding/binary"

	"github.com/cosmonaut/blog/x/blog/types"
	"github.com/cosmos/cosmos-sdk/store/prefix"
	sdk "github.com/cosmos/cosmos-sdk/types"
	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 )

 func (k msgServer) DeleteComment(goCtx context.Context, msg *types.MsgDeleteComment) (*types.MsgDeleteCommentResponse, error) {
	ctx := sdk.UnwrapSDKContext(goCtx)

	comment, exist := k.GetComment(ctx, msg.Id)
	if !exist {
		return nil, sdkerrors.Wrapf(types.ErrID, "Comment doesnt exist")
	}

	if msg.PostID != comment.PostID {
		return nil, sdkerrors.Wrapf(types.ErrID, "Post Blog Id does not exist for which comment with Blog Id %d was made", msg.PostID)
	}

	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.CommentKey))
	bz := make([]byte, 8)
	binary.BigEndian.PutUint64(bz, comment.Id)
	store.Delete(bz)

	return &types.MsgDeleteCommentResponse{}, nil
}

```

## Display posts

Implement logic to query existing posts:

```bash
starport scaffold query comments id:uint --response title,body
```

Also in `proto/blog/query.proto`, make these updates:

```go

> message QueryCommentsRequest {

Reviewer: ```suggestion
"blog/post.proto";

message QueryCommentsRequest {
```

To resolve this error:

```
cannot build app:

        cannot build app:

        error while running command /var/folders/8t/_76n_22x4_1ffzqrttf29yyc0000gn/T/protoc831205405 -I /var/folders/8t/_76n_22x4_1ffzqrttf29yyc0000gn/T/1510796494 -I /Users/fadeev/blog/proto -I /Users/fadeev/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.44.5/proto -I /Users/fadeev/go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.44.5/third_party/proto --gocosmos_out=plugins=interfacetype+grpc,Mgoogle/protobuf/any.proto=github.com/cosmos/cosmos-sdk/codec/types:. /Users/fadeev/blog/proto/blog/comment.proto /Users/fadeev/blog/proto/blog/genesis.proto /Users/fadeev/blog/proto/blog/params.proto /Users/fadeev/blog/proto/blog/post.proto /Users/fadeev/blog/proto/blog/query.proto /Users/fadeev/blog/proto/blog/tx.proto: blog/query.proto:70:3: "cosmonaut.blog.blog.Post" seems to be defined in "blog/post.proto", which is not imported by "blog/query.proto".  To use it here, please add the necessary import.
: exit status 1
```

# Add Associated Comments to a Blog Post

In this tutorial, you create a new message to add comments to a blog post. These tutorial steps 

By completing this tutorial, you will learn about:

* Scaffolding a new `list` with proto functions and keeper functions
* Adding comments to existing blog posts
* Querying for blog posts that have associated comments
* Deleting comments from a blog post
* Implementing logic for writing comments to the blockchain

**Note:** For this tutorial, adding comments is available only to blog posts that are no older than 100 blocks. The 100 block value has been hard coded for rapid testing. You can increase the block count to a larger number to achieve a longer time period before commenting is disabled.

## Prerequisites

- This tutorial is an extension of and requires completion of the [Module Basics: Build a Blog](index.md) tutorial. 

## Core Concepts 

This tutorial relies on the `blog` blockchain that you built in the build a blog tutorial. 

To get the useful functions for this tutorial, you use the `starport scaffold list comment --no-message` command. Make sure to familiarize yourself with the command.

> ## Fetch Functions Using List Command

Reviewer 2: @alijnmerchant21  be sure to review the tutorial template, edit the file to see all of the stylistic guidelines https://github.com/tendermint/starport/edit/develop/docs/contributing/templates/tutorial-template.md 

For example 

> `<!-- Use Title Case for all Titles, see https://capitalizemytitle.com/ -->`

> 1. Navigate to the `blog` directory that you created in the build a blog tutorial.

Reviewer 2: moved this step here so the actions are together
